### ISSUE TITLE
flaskwebsocket Warning: engineio:Invalid session

### ISSUE THREAD

**Issue by zwj2017-NK** [miguelgrinberg/Flask-SocketIO#591]
server：
2017-11-24 16:26:23,657:WARNING:engineio:Invalid session e9b2c478cae44f4faa6fa8b7465deff5
2017-11-24 16:26:24,654:WARNING:engineio:Invalid session e9b2c478cae44f4faa6fa8b7465deff5
2017-11-24 16:26:26,662:WARNING:engineio:Invalid session 95900ab0219a49b0a06a67213e868a94
2017-11-24 16:26:26,664:WARNING:engineio:Invalid session 95900ab0219a49b0a06a67213e868a94
2017-11-24 16:26:27,654:WARNING:engineio:Invalid session 95900ab0219a49b0a06a67213e868a94
2017-11-24 16:26:28,671:WARNING:engineio:Invalid session b52f6520cb4c4e3a81385e121f772ae3
2017-11-24 16:26:31,654:INFO:engineio:1a76fa9c9977426ca816de51cde77dfc: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': [], 'sid': '1a76fa9c9977426ca816de51cde77dfc'}
2017-11-24 16:26:31,654:INFO:engineio:1a76fa9c9977426ca816de51cde77dfc: Sending packet MESSAGE data 0
2017-11-24 16:26:31,664:INFO:engineio:1a76fa9c9977426ca816de51cde77dfc: Received packet MESSAGE data 0/test
2017-11-24 16:26:31,665:INFO:engineio:1a76fa9c9977426ca816de51cde77dfc: Sending packet MESSAGE data 0/test
2017-11-24 16:26:32,657:INFO:engineio:1a76fa9c9977426ca816de51cde77dfc: Received packet CLOSE data None
2017-11-24 16:26:33,654:INFO:engineio:0d6a148655884cec9eed310c7d816f2e: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': [], 'sid': '0d6a148655884cec9eed310c7d816f2e'}
2017-11-24 16:26:33,654:INFO:engineio:0d6a148655884cec9eed310c7d816f2e: Sending packet MESSAGE data 0
2017-11-24 16:26:33,664:INFO:engineio:0d6a148655884cec9eed310c7d816f2e: Received packet MESSAGE data 0/test
2017-11-24 16:26:33,664:INFO:engineio:0d6a148655884cec9eed310c7d816f2e: Sending packet MESSAGE data 0/test
2017-11-24 16:26:37,664:WARNING:engineio:Invalid session 80dd2c2c3fb8409e8386b8138fa10543
2017-11-24 16:26:38,654:WARNING:engineio:Invalid session 80dd2c2c3fb8409e8386b8138fa10543
2017-11-24 16:26:40,654:INFO:engineio:22fa7380dfc54dedb8457da8b29ed448: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': [], 'sid': '22fa7380dfc54dedb8457da8b29ed448'}
2017-11-24 16:26:40,654:INFO:engineio:22fa7380dfc54dedb8457da8b29ed448: Sending packet MESSAGE data 0
2017-11-24 16:26:40,663:INFO:engineio:22fa7380dfc54dedb8457da8b29ed448: Received packet MESSAGE data 0/test
2017-11-24 16:26:40,663:INFO:engineio:22fa7380dfc54dedb8457da8b29ed448: Sending packet MESSAGE data 0/test
2017-11-24 16:26:54,662:INFO:engineio:06ef622918af4ca480d7e5e47180f500: Client is gone, closing socket
[2017-11-24 16:26:54 +0000] [4607] [ERROR] Error handling request /socket.io/?EIO=3&transport=polling&t=1511511954660-77&sid=06ef622918af4ca480d7e5e47180f500
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/sync.py", line 135, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/sync.py", line 176, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/flask_socketio/__init__.py", line 42, in __call__
    start_response)
  File "/usr/local/lib/python2.7/dist-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/socketio/server.py", line 360, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 268, in handle_request
    self.disconnect(sid)
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 202, in disconnect
    self._get_socket(sid).close()
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 408, in _get_socket
    raise KeyError('Session is disconnected')
KeyError: 'Session is disconnected'

Front end：

socket.io/?EIO=3&transport=polling&t=1511511881487-1&sid=5135fbf01d7c4e5fa2bce4c2969e61f3 Failed to load resource: the server responded with a status of 400 (BAD REQUEST)
socket.io/?EIO=3&transport=polling&t=1511511881487-2&sid=5135fbf01d7c4e5fa2bce4c2969e61f3 Failed to load resource: the server responded with a status of 400 (BAD REQUEST)
socket.io/?EIO=3&transport=polling&t=1511511881492-3&sid=5135fbf01d7c4e5fa2bce4c2969e61f3 Failed to load resource: the server responded with a status of 400 (BAD REQUEST)

After upgrade python-engineio to 1.6.0. The first run, the histogram is displayed.Not too long ago, the log shows that the connection has been disconnected.Run again, the picture is not displayed. I have developed an application that enables real-time interaction between the server and the front end to pass data.I want to add this application to another application(example:superset), but I did not succeed.

**Comment by zwj2017-NK** [comment]
@miguelgrinberg Thank you for your helping. I look forward to hearing from you.

**Comment by zwj2017-NK** [comment]
**The following is a normal log output, but unfortunately the good news is not long：**

2017-11-24 16:52:37,880:INFO:socketio:received event "my event" from f9d31e8cda074b7a9b51307afeb6c3c9 [/test]
connected!:1511513488973
2017-11-24 16:52:37,883:INFO:socketio:emitting event "connect" to f9d31e8cda074b7a9b51307afeb6c3c9 [/test]
2017-11-24 16:52:37,883:INFO:engineio:f9d31e8cda074b7a9b51307afeb6c3c9: Sending packet MESSAGE data 2/test,["connect","I am server!"]
2017-11-24 16:52:40,733:INFO:engineio:f9d31e8cda074b7a9b51307afeb6c3c9: Received packet MESSAGE data 2/test,["my chart","Charts!"]
2017-11-24 16:52:40,733:INFO:socketio:received event "my chart" from f9d31e8cda074b7a9b51307afeb6c3c9 [/test]
Charts!:1511513488973
Charts! 7 next 4
2017-11-24 16:52:40,734:INFO:socketio:emitting event "my chart" to f9d31e8cda074b7a9b51307afeb6c3c9 [/test]
2017-11-24 16:52:40,734:INFO:engineio:f9d31e8cda074b7a9b51307afeb6c3c9: Sending packet MESSAGE data 2/test,["my chart","{\"data\": [{\"num\": 100, \"name\": \"allpe\"}, {\"num\": 123, \"name\": \"peach\"}, {\"num\": 234, \"name\": \"Pear\"}, {\"num\": 20, \"name\": \"avocado\"}, {\"num\": 30, \"name\": \"cantaloupe\"}, {\"num\": 77, \"name\": \"Banana\"}, {\"num\": 43, \"name\": \"Grape\"}, {\"num\": 22, \"name\": \"apricot\"}]}"]

**Comment by miguelgrinberg** [comment]
How many server processes do you have? Just one or more?

**Comment by zwj2017-NK** [comment]
My application is placed directly on the other, that is superset. I also tried to modify the number of superset's workers, changeing from default 2 to 1. But the problem remains. Other superset code I have not changed. Thank you for your replying

**my views.py:**

app.config['SECRET_KEY'] = 'secret'
socketio = SocketIO(app, logger=True, engineio_logger=True)
app.config['SECRET_KEY'] = 'secret'
t = time.time()
nowTime = lambda: int(round(t * 1000))
data_chart = {
        "data": [
            {"name": "allpe", "num": 100},
            {"name": "peach", "num": 123},
            {"name": "Pear", "num": 234},
            {"name": "avocado", "num": 20},
            {"name": "cantaloupe", "num": 30},
            {"name": "Banana", "num": 77},
            {"name": "Grape", "num": 43},
            {"name": "apricot", "num": 22}
        ]
    }
name = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']

@app.route('/echarts')
def index():
    return render_template('test.html')

@socketio.on('my event', namespace='/test')
def handle_my_custom_event(message):
    print (message['data'] + ':' + str(nowTime()))
    emit('connect', "I am server!")

@socketio.on('my chart', namespace='/test')
def handle_json(message):
    print (message + ':' + str(nowTime()))
    msg = str(message)
    print (msg, len(msg), 'next', len('next'))
    list_add = {}
    if msg == 'next':
        i = random.randint(0, 25)
        j = random.randint(0, 25)
        new_name = name[i] + name[j]
        list_add['name'] = new_name
        list_add['num'] = random.randint(0, 500)

        data_chart['data'].append(list_add)
        data_chart['data'].pop(0)
        emit('my chart', json.dumps(data_chart))
    else:
        emit('my chart', json.dumps(data_chart))


**my test.html:**


<script type="text/javascript" charset="utf-8">  
    var namespace = '/test'  
    var socket = io.connect('http://' + document.domain + ':' + location.port + namespace);  
    socket.on('connect', function(msg)//对connect事件进行监听  
    {  
       var timestamp = Date.parse(new Date());  
       if(msg)  
       {  
          //alert(msg+timestamp);  
          //socket.emit('my json','I am client!')  
          socket.emit('my chart', 'Charts!')  
       }  
    });  

    /*
    socket.on('my json', function(msg)
    {
       alert(msg);
    });
    */

    function a()
    {
        //向my event事件上发送信息
        socket.emit('my event', {data: 'connected!'})
    }

    setTimeout(a, 1000);

    //echarts画图
    var a = echarts;
    var myChart = a.init(document.getElementById('main'));

    // 显示标题，图例和空的坐标轴

    option={
        title:{
            text: '爬虫今日抓取数据图'
        },
        tooltip:{
            trigger: 'axis'
        },
        legend: {
            data:['今日数据']
        },
        toolbox: {
            show : true,
            feature : {
                mark : {show: true},
                dataView : {show: true, readOnly: false},
                magicType : {show: true, type: ['line', 'bar']},
                restore : {show: true},//显示原来的图
                saveAsImage : {show: true}//保存图片
            }
        },
        calculable : true,

        xAxis:[
        {
            type : 'category',
            boundaryGap : false,
            data : []
        }],
        yAxis:[
        {
            type : 'value',
            axisLabel:{
               formatter: '{value}'
            }
        }],
        series:[
        {
             name:'最多数量',
             type:'line',
             //showSymbol: false,
             //hoverAnimation: false,
             data:[],
             markPoint:{
                 data:[
                     {type : 'max', name: '最大值'},
                     {type : 'min', name: '最小值'}
                     ]
             },
             markLine : {
                 data : [
                     {type : 'average', name: '平均值'}
                 ]
             }
         },]
    };


    myChart.setOption(option,true);
    myChart.showLoading();    //数据加载完之前先显示一段简单的loading动画
    var names=[];    //名称数组（实际用来盛放X轴坐标值）
    var nums=[];    //数量数组（实际用来盛放Y坐标值）

    socket.on('my chart', function b(str)
    {
         //请求成功时执行该函数内容，result即为服务器返回的json对象
        var result = JSON.parse(str);
        names.splice(0,names.length);
        nums.splice(0,nums.length);

        if (result)
        {
            for(var i=0;i<result["data"].length;i++)//alert(result["data"][i]["name"]);
            {
                names.push(result["data"][i]["name"]);    //挨个取出名称并填入类别数组
            }
            for(var i=0;i<result["data"].length;i++)// alert(result["data"][i]["num"]);
            {
                nums.push(result["data"][i]["num"]);    //挨个取出数量并填入销量数组
            }
            myChart.hideLoading();    //隐藏加载动画

            option_data={        //加载数据图表
                xAxis: {
                     data: names
                },
                series: [{// 根据名字对应到相应的系列
                     name: '数量',
                     data: nums
               }]
            }
            myChart.setOption(option_data);
        }

        function d(){
             socket.emit('my chart', 'next');
        };
        setTimeout(d, 5000);
    });
</script>

**Comment by miguelgrinberg** [comment]
@zwj2017-NK Sorry, I don't understand your answer. What's "superset"?

If you have more than one worker, then your load balancer must use sticky sessions. Can you show me how do you start your server?

**Comment by zwj2017-NK** [comment]
**I use "superset runserver" command to start the server. I'm so sorry, superset is designed to be highly available. It is “cloud-native” as it has been designed scale out in large, distributed environments, and works well inside containers. From flask to FAB, to superset, is a gradual process.**


cli.py
#!/usr/bin/env python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import logging
from datetime import datetime
from subprocess import Popen

from colorama import Fore, Style
from flask_migrate import MigrateCommand
from flask_script import Manager
from flask_socketio import SocketIO

from superset import app, db, security, utils




config = app.config
celery_app = utils.get_celery_app(config)

manager = Manager(app)
manager.add_command('db', MigrateCommand)


#
# app.config['SECRET_KEY'] = 'secret'
# socketio = SocketIO(app, logger=True, engineio_logger=True)
#

@manager.command
def init():
    """Inits the Superset application"""
    security.sync_role_definitions()


@manager.option(
    '-d', '--debug', action='store_true',
    help="Start the web server in debug mode")
@manager.option(
    '-n', '--no-reload', action='store_false', dest='no_reload',
    default=config.get("FLASK_USE_RELOAD"),
    help="Don't use the reloader in debug mode")
@manager.option(
    '-a', '--address', default=config.get("SUPERSET_WEBSERVER_ADDRESS"),
    help="Specify the address to which to bind the web server")
@manager.option(
    '-p', '--port', default=config.get("SUPERSET_WEBSERVER_PORT"),
    help="Specify the port on which to run the web server")
@manager.option(
    '-w', '--workers',
    default=config.get("SUPERSET_WORKERS", 1),
    help="Number of gunicorn web server workers to fire up")
@manager.option(
    '-t', '--timeout', default=config.get("SUPERSET_WEBSERVER_TIMEOUT"),
    help="Specify the timeout (seconds) for the gunicorn web server")
@manager.option(
    '-s', '--socket', default=config.get("SUPERSET_WEBSERVER_SOCKET"),
    help="Path to a UNIX socket as an alternative to address:port, e.g. "
         "/var/run/superset.sock. "
         "Will override the address and port values.")
def runserver(debug, no_reload, address, port, timeout, workers, socket):
    """Starts a Superset web server."""
    debug = debug or config.get("DEBUG")
    if debug:
        print(Fore.BLUE + '-=' * 20)
        print(
            Fore.YELLOW + "Starting Superset server in " +
            Fore.RED + "DEBUG" +
            Fore.YELLOW + " mode")
        print(Fore.BLUE + '-=' * 20)
        print(Style.RESET_ALL)
        # app.run()
        app.run(
            host='0.0.0.0',
            port=int(port),
            threaded=True,
            debug=True,
            use_reloader=no_reload)
    else:
        addr_str = " unix:{socket} " if socket else" {address}:{port} "
        cmd = (
            "gunicorn "
            "-w {workers} "
            "--timeout {timeout} "
            "-b " + addr_str +
            "--limit-request-line 0 "
            "--limit-request-field_size 0 "
            "superset:app").format(**locals())
        print(Fore.GREEN + "Starting server with command: ")
        print(Fore.YELLOW + cmd)
        print(Style.RESET_ALL)
        Popen(cmd, shell=True).wait()


@manager.option(
    '-v', '--verbose', action='store_true',
    help="Show extra information")
def version(verbose):
    """Prints the current version number"""
    print(Fore.BLUE + '-=' * 15)
    print(Fore.YELLOW + "Superset " + Fore.CYAN + "{version}".format(
        version=config.get('VERSION_STRING')))
    print(Fore.BLUE + '-=' * 15)
    if verbose:
        print("[DB] : " + "{}".format(db.engine))
    print(Style.RESET_ALL)


@manager.option(
    '-t', '--load-test-data', action='store_true',
    help="Load additional test data")
def load_examples(load_test_data):
    """Loads a set of Slices and Dashboards and a supporting dataset """
    from superset import data
    print("Loading examples into {}".format(db))

    data.load_css_templates()

    print("Loading energy related dataset")
    data.load_energy()

    print("Loading [World Bank's Health Nutrition and Population Stats]")
    data.load_world_bank_health_n_pop()

    print("Loading [Birth names]")
    data.load_birth_names()

    print("Loading [Random time series data]")
    data.load_random_time_series_data()

    print("Loading [Random long/lat data]")
    data.load_long_lat_data()

    print("Loading [Country Map data]")
    data.load_country_map_data()

    print("Loading [Multiformat time series]")
    data.load_multiformat_time_series_data()

    print("Loading [Misc Charts] dashboard")
    data.load_misc_dashboard()

    print("Loading [My Conn And Dns data]")
    data.load_test_pcap_data()

    print("Loading [My Csv data]")
    data.load_test_csv_data()

    print("Loading [My test map data]")
    data.load_test_map_data()

    if load_test_data:
        print("Loading [Unicode test data]")
        data.load_unicode_test_data()


@manager.option(
    '-d', '--datasource',
    help=(
            "Specify which datasource name to load, if omitted, all "
            "datasources will be refreshed"))
@manager.option(
    '-m', '--merge',
    help=(
            "Specify using 'merge' property during operation. "
            "Default value is False "))
def refresh_druid(datasource, merge):
    """Refresh druid datasources"""
    session = db.session()
    from superset.connectors.druid.models import DruidCluster
    for cluster in session.query(DruidCluster).all():
        try:
            cluster.refresh_datasources(datasource_name=datasource,
                                        merge_flag=merge)
        except Exception as e:
            print(
                "Error while processing cluster '{}'\n{}".format(
                    cluster, str(e)))
            logging.exception(e)
        cluster.metadata_last_refreshed = datetime.now()
        print(
            "Refreshed metadata from cluster "
            "[" + cluster.cluster_name + "]")
    session.commit()


@manager.command
def update_datasources_cache():
    """Refresh sqllab datasources cache"""
    from superset.models.core import Database
    for database in db.session.query(Database).all():
        print('Fetching {} datasources ...'.format(database.name))
        try:
            database.all_table_names(force=True)
            database.all_view_names(force=True)
        except Exception as e:
            print('{}'.format(e.message))


@manager.option(
    '-w', '--workers',
    type=int,
    help="Number of celery server workers to fire up")
def worker(workers):
    """Starts a Superset worker for async SQL query execution."""
    if workers:
        celery_app.conf.update(CELERYD_CONCURRENCY=workers)
    elif config.get("SUPERSET_CELERY_WORKERS"):
        celery_app.conf.update(
            CELERYD_CONCURRENCY=config.get("SUPERSET_CELERY_WORKERS"))

    worker = celery_app.Worker(optimization='fair')
    worker.start()


@manager.option(
    '-p', '--port',
    default='5555',
    help=('Port on which to start the Flower process'))
@manager.option(
    '-a', '--address',
    default='localhost',
    help=('Address on which to run the service'))
def flower(port, address):
    """Runs a Celery Flower web server

    Celery Flower is a UI to monitor the Celery operation on a given
    broker"""
    BROKER_URL = celery_app.conf.BROKER_URL
    cmd = (
        "celery flower "
        "--broker={BROKER_URL} "
        "--port={port} "
        "--address={address} "
    ).format(**locals())
    print(Fore.GREEN + "Starting a Celery Flower instance")
    print(Fore.BLUE + '-=' * 40)
    print(Fore.YELLOW + cmd)
    print(Fore.BLUE + '-=' * 40)
    Popen(cmd, shell=True).wait()

__init__.py
"""Package's main module!"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import logging
from logging.handlers import TimedRotatingFileHandler

import json
import os

from flask import Flask, redirect
from flask_appbuilder import SQLA, AppBuilder, IndexView
from flask_appbuilder.baseviews import expose
from flask_migrate import Migrate
from flask_wtf.csrf import CSRFProtect
from werkzeug.contrib.fixers import ProxyFix

from superset.connectors.connector_registry import ConnectorRegistry
from superset import utils, config  # noqa

APP_DIR = os.path.dirname(__file__)
CONFIG_MODULE = os.environ.get('SUPERSET_CONFIG', 'superset.config')

with open(APP_DIR + '/static/assets/backendSync.json', 'r') as f:
    frontend_config = json.load(f)

app = Flask(__name__)
app.config.from_object(CONFIG_MODULE)
conf = app.config


from flask_socketio import SocketIO
app.config['SECRET_KEY'] = 'secret'
socketio = SocketIO(app, logger=True, engineio_logger=True) #async_mode=None


#################################################################
# Handling manifest file logic at app start
#################################################################
MANIFEST_FILE = APP_DIR + '/static/assets/dist/manifest.json'
manifest = {}


def parse_manifest_json():
    global manifest
    try:
        with open(MANIFEST_FILE, 'r') as f:
            manifest = json.load(f)
    except Exception:
        print("no manifest file found at " + MANIFEST_FILE)


def get_manifest_file(filename):
    if app.debug:
        parse_manifest_json()
    return '/static/assets/dist/' + manifest.get(filename, '')


parse_manifest_json()


@app.context_processor
def get_js_manifest():
    return dict(js_manifest=get_manifest_file)


#################################################################

for bp in conf.get('BLUEPRINTS'):
    try:
        print("Registering blueprint: '{}'".format(bp.name))
        app.register_blueprint(bp)
    except Exception as e:
        print("blueprint registration failed")
        logging.exception(e)

if conf.get('SILENCE_FAB'):
    logging.getLogger('flask_appbuilder').setLevel(logging.ERROR)

if not app.debug:
    # In production mode, add log handler to sys.stderr.
    app.logger.addHandler(logging.StreamHandler())
    app.logger.setLevel(logging.INFO)
logging.getLogger('pyhive.presto').setLevel(logging.INFO)

db = SQLA(app)

if conf.get('WTF_CSRF_ENABLED'):
    csrf = CSRFProtect(app)
    csrf_exempt_list = conf.get('WTF_CSRF_EXEMPT_LIST', [])
    for ex in csrf_exempt_list:
        csrf.exempt(ex)

utils.pessimistic_connection_handling(db.engine)

cache = utils.setup_cache(app, conf.get('CACHE_CONFIG'))
tables_cache = utils.setup_cache(app, conf.get('TABLE_NAMES_CACHE_CONFIG'))

migrate = Migrate(app, db, directory=APP_DIR + "/migrations")

# Logging configuration
logging.basicConfig(format=app.config.get('LOG_FORMAT'))
logging.getLogger().setLevel(app.config.get('LOG_LEVEL'))

if app.config.get('ENABLE_TIME_ROTATE'):
    logging.getLogger().setLevel(app.config.get('TIME_ROTATE_LOG_LEVEL'))
    handler = TimedRotatingFileHandler(
        app.config.get('FILENAME'),
        when=app.config.get('ROLLOVER'),
        interval=app.config.get('INTERVAL'),
        backupCount=app.config.get('BACKUP_COUNT'))
    logging.getLogger().addHandler(handler)

if app.config.get('ENABLE_CORS'):
    from flask_cors import CORS
    CORS(app, **app.config.get('CORS_OPTIONS'))

if app.config.get('ENABLE_PROXY_FIX'):
    app.wsgi_app = ProxyFix(app.wsgi_app)

if app.config.get('ENABLE_CHUNK_ENCODING'):

    class ChunkedEncodingFix(object):
        def __init__(self, app):
            self.app = app

        def __call__(self, environ, start_response):
            # Setting wsgi.input_terminated tells werkzeug.wsgi to ignore
            # content-length and read the stream till the end.
            if environ.get('HTTP_TRANSFER_ENCODING', '').lower() == u'chunked':
                environ['wsgi.input_terminated'] = True
            return self.app(environ, start_response)

    app.wsgi_app = ChunkedEncodingFix(app.wsgi_app)

if app.config.get('UPLOAD_FOLDER'):
    try:
        os.makedirs(app.config.get('UPLOAD_FOLDER'))
    except OSError:
        pass

for middleware in app.config.get('ADDITIONAL_MIDDLEWARE'):
    app.wsgi_app = middleware(app.wsgi_app)


class MyIndexView(IndexView):
    @expose('/')
    def index(self):
        return redirect('/superset/welcome')


appbuilder = AppBuilder(
    app,
    db.session,
    base_template='superset/base.html',
    indexview=MyIndexView,
    security_manager_class=app.config.get("CUSTOM_SECURITY_MANAGER"))

sm = appbuilder.sm

results_backend = app.config.get("RESULTS_BACKEND")

# Registering sources
module_datasource_map = app.config.get("DEFAULT_MODULE_DS_MAP")
module_datasource_map.update(app.config.get("ADDITIONAL_MODULE_DS_MAP"))
ConnectorRegistry.register_sources(module_datasource_map)

from superset import views  # noqa

**Comment by miguelgrinberg** [comment]
@zwj2017-NK Did you read the documentation regarding how to start a Socket.IO server with gunicorn? You are not using a valid gunicorn command, you need to scale down to a single worker, and also use either eventlet or gevent.

**Comment by zwj2017-NK** [comment]
**I kown start Socket.IO server using socketio.run(app) command. I now have two workers, and used gevent. I also tried using socketio.run (app) to turn on the server, but the same thing happened. Thank you for staying up late to answer my questions, thank you**

**Comment by miguelgrinberg** [comment]
You can't have two workers. If you want to use multiple workers, then you need to use a more complicated set up using a message queue. Please see the documentation, all the deployment options for your server are listed there.

**Comment by zwj2017-NK** [comment]
I do not know how to combine superset and Echarts. I put Echarts source code into my github account, the connection address is as follows.I recently learned your Chinese version of the book of Flask web development, I really like this book.Can you squeeze the time to help me achieve superset and echarts combination? I know this request is very rude,but I really do not know how to solve this problem, and if you can help me, I would appreciate it
**https://github.com/zwj2017-NK/Superset-Echarts.git**

**Comment by miguelgrinberg** [comment]
@zwj2017-NK I have no idea what superset is or how it works. Did you try Stack Overflow?

**Comment by zwj2017-NK** [comment]
Superset is a data visualization tool based on the application of flask-appbuilder, Apache's open source tool.I have not tried Stack Overflow to solve the problem. My idea is very simple, is to add a connection in the superset web page to achieve server and front-end data interaction.This question should be very simple for you. Can you just try?

**Comment by miguelgrinberg** [comment]
This issue will be automatically closed due to being inactive for more than six months. Seeing that I haven't responded to your last comment, it is quite possible that I have dropped the ball on this issue and I apologize about that. If that is the case, do not take the closing of the issue personally as it is an automated process doing it, just reopen it and I'll get back to you.
